torchdata/datapipes/iter/util/prefetch.py

```diff
@@ -0,0 +1,209 @@
+        super().__init__(f"Fail to fetch data within {timeout} seconds")
+
+
+class _EndOfPrefetch:
+    ...
+
+
+@dataclass
+class Expected:
+    r"""
+    Expected data provided to callback function in ``_PrefetchExecutor``.
+    """
+    index: int
+    error: Optional[BaseException] = None
+
+    def has_error(self) -> bool:
+        return self.error is not None
+
+
+class _PrefetchExecutor:
+    def __init__(
+        self,
+        datapipe_iterator: Iterator,
+        prefetch_size: int = 1,
+        callback_fn: Optional[Callable[[Expected], None]] = None,
+        timeout: int = default_timeout_in_s,
+    ) -> None:
+        self.datapipe_iterator = datapipe_iterator
+        self.prefetch_size = prefetch_size
+        self.callback_fn = callback_fn
+        self.timeout = timeout
+        # Use max_workers as 1 to guarantee the order of data fetched from iterator
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._futures: Deque[Future] = deque()
+        self._lock = threading.RLock()
+        self._end_flag = False
+        self._idx = 0
+        for _ in range(prefetch_size):
+            with self._lock:
+                if self._end_flag:
+                    break
+            fetch_future: Future = self._executor.submit(self.fetch_next)
+            fetch_future.add_done_callback(partial(self._done_callback_fn, self._idx))
+            self._futures.append(fetch_future)
+            with self._lock:
+                self._idx += 1
+
+    def fetch_next(self):
+        return next(self.datapipe_iterator)
+
+    def _done_callback_fn(self, index: int, f: Future):
+        if f.exception():
+            with self._lock:
+                self._end_flag = True
+        if self.callback_fn is not None:
+            self._executor.submit(self.callback_fn, Expected(index, f.exception()))
+
+    def return_next(self):
+        if self._futures:
+            fetch_future = self._futures.popleft()
+            try:
+                data = fetch_future.result(timeout=self.timeout)
+            except TimeoutError:
+                raise PrefetchTimeoutError(self.timeout)
+            with self._lock:
+                if not self._end_flag:
+                    next_future = self._executor.submit(self.fetch_next)
+                    next_future.add_done_callback(partial(self._done_callback_fn, self._idx))
+                    self._futures.append(next_future)
+                    self._idx += 1
+        else:
+            data = _EndOfPrefetch()
+        return data
+
+    def shutdown(self):
+        self._executor.shutdown(wait=True)
+
+
+@functional_datapipe("fullsync")
+class FullSyncIterDataPipe(IterDataPipe[T_co]):
+    r"""
+    Synchronizes data across distributed processes to prevent hanging during training,
+    which is caused by uneven sharded data (functional name: ``fullsync``). It should
+    be appended at the end of the graph of ``DataPipe`` by ``DistributedReadingService``
+    automatically.
+
```

Comment:
Question: do we recommend against usage of this DataPipe outside of a ReadingService? If not, can we potentially include an example?

Reply:
Makes sense. Will add it even though we should always recommend users relying on RS